proptest-state-machine/src/iterative_runner.rs

```diff
@@ -0,0 +1,861 @@
+use std::{cell::RefCell, marker::PhantomData};
+
+use proptest::{
+    bits::{BitSetLike, VarBitSet},
+    collection::SizeRange,
+    num::sample_uniform_incl,
+    prelude::Just,
+    std_facade::fmt::Debug,
+    strategy::{BoxedStrategy, Strategy, ValueTree},
+    test_runner::{Config, Reason, TestError, TestRunner},
+};
+
+pub fn test_sequential<
+    T,
+    MS,
+    SS,
+    M: IterativeModelStateMachine<ModelState = MS, Transition = T, SutState = SS>,
+    S: IterativeSutStateMachine<SutState = SS, Transition = T, ModelState = MS>,
+>(
+    config: Config,
+    mut mod_state: MS,
+    mut sut_state: SS,
+    transitions: Vec<T>,
+) where
+    T: Clone + Debug,
+{
+    let trans_len = transitions.len();
+    #[cfg(feature = "std")]
+    if config.verbose >= super::INFO_LOG {
+        eprintln!();
+        eprintln!("Running a test case with {} transitions.", trans_len);
+    }
+    #[cfg(not(feature = "std"))]
+    let _ = (config, trans_len);
+
+    // Check the invariants on the initial state
+    S::check_invariants(&sut_state, &mod_state);
+
+    for (ix, transition) in transitions.into_iter().enumerate() {
+        #[cfg(feature = "std")]
+        if config.verbose >= super::INFO_LOG {
+            eprintln!();
+            eprintln!(
+                "Applying transition {}/{}: {:?}",
+                ix + 1,
+                trans_len,
+                transition
+            );
+        }
+        #[cfg(not(feature = "std"))]
+        let _ = ix;
+        assert_eq!(
+            true,
+            M::preconditions(&mod_state, &sut_state, &transition),
+            "precondition failed on transition: {:?}",
+            transition
+        );
+        // Apply the transition on the states
+        mod_state = M::apply(mod_state, &transition);
+        sut_state = S::apply(sut_state, &mod_state, transition);
+
+        // Check the invariants after the transition is applied
+        S::check_invariants(&sut_state, &mod_state);
+    }
+
+    S::teardown(sut_state, mod_state)
+}
+
+pub trait IterativeModelStateMachine {
```

Comment:
The model in the IterativeRunner is an entity that supports generating of the test scenarios rather than the reference model. If the idea is to have both IterativeRunner and the current SequentialRunner in the crate I would say it might be desired to have explicit distinction between the two.